### DISCUSSION TITLE
How to call a hook on closing the window / application exit

### DISCUSSION THREAD

**Discussion by Zajozor** [slint-ui/slint#943] [Q&A] [answered]
Imagine an application which has a state. I want to persist the state upon closing a window/application so that I can load it again the next time it starts. (writing the state to disk on every change is not viable)
**How do I create some sort of a hook which runs on closing a window?**

I did not find any examples that would help. Can you help me?

I tried something like
```
impl Drop for MainWindow {
    fn drop(&mut self) {
        eprintln!("Doing some final cleanup");
    }
}
```
But that does not really work:
```
error[E0509]: cannot move out of type `MainWindow`, which implements the `Drop` trait
    --> .../out/main.rs:2348:14
     |
2348 |              value . 0 }
     |              ^^^^^^^^^
     |              |
     |              cannot move out of here
     |              move occurs because `value.0` has type `slint::re_exports::VRc<slint::re_exports::ComponentVTable, InnerMainWindow>`, which does not implement the `Copy` trait
```

I could also go with some solutions which are not tied to the GUI, but it feels like a workaround.
Also "do something when this window closes" is not such an uncommon request 😅 

Thanks! :)

**Comment by ogoffart** [comment]
We currently do not have such way.
We were discussing a bit in this issue: https://github.com/slint-ui/slint/issues/325 to add a callback to the Window.

What kind of cleanup exactly do you need to perform?

**Reply by Zajozor** [reply]
I wanted to save the application state before exiting.

Imagine you are creating a clone of MS Word. And the logic you want to implement is "before you exit, save the file".
I'd like to create some "onexit hook" which would do that.

Is there some other recommended way to achieve this?
the `close_event` you talk about in #325 sounds like that..

**Comment by ogoffart** [comment, marked as answer]
In the master branch, we added a on_close_requested function to the Window: https://slint-ui.com/snapshots/master/docs/rust/slint/struct.window#method.on_close_requested

**Reply by Zajozor** [reply]
That's amazing! good work! :)

**Comment by DizzyHSlightlyVoided** [comment]
How do you set this in Node?
